pkg/scheduler/plugins/gang/plugin.go

```diff
@@ -0,0 +1,519 @@
+	waitTime, s := p.PermitCheck(pod)
+	var retStatus *framework.Status
+	switch s {
+	case NotFoundInCache:
+		return framework.NewStatus(framework.Unschedulable, "Gang not found in gangCache"), 0
+	case Wait:
+		klog.Infof("Pod: %v from gang: %v is waiting to be scheduled at Permit stage",
+			getNamespaceSplicingName(pod.Namespace, pod.Name),
+			getNamespaceSplicingName(pod.Namespace,
+				getGangNameByPod(pod)))
+		retStatus = framework.NewStatus(framework.Wait)
+		p.ActivateGang(pod, state)
+	case Success:
+		p.AllowGangGroup(pod)
+		retStatus = framework.NewStatus(framework.Success)
+		waitTime = 0
+	}
+	return retStatus, waitTime
+}
+
+func (p *GangPlugin) PermitCheck(pod *corev1.Pod) (time.Duration, Status) {
+	gangName := getGangNameByPod(pod)
+	if gangName == "" {
+		return 0, Success
+	}
+	gangId := getNamespaceSplicingName(pod.Namespace,
+		gangName)
+	gang := p.gangCache.getGangFromCache(gangId)
+	if gang == nil {
+		klog.Errorf("can't find gang, gangName:%v, podName:%v", gangId,
+			getNamespaceSplicingName(pod.Namespace, pod.Name))
+		return 0, NotFoundInCache
+	}
+
+	gangGroup := gang.getGangGroup()
+	allGangGroupSatisfied := true
+	// only the gang itself
+	if len(gangGroup) == 0 {
+		allGangGroupSatisfied = gang.isGangResourceSatisfied()
+	} else {
+		// check each gang group
+		for _, groupName := range gangGroup {
+			gangTmp := p.gangCache.getGangFromCache(groupName)
+			if gangTmp != nil {
+				if !gangTmp.isGangResourceSatisfied() {
+					allGangGroupSatisfied = false
+					break
+				}
+			}
+		}
+	}
+	if !allGangGroupSatisfied {
+		gang.setTimeoutStartTime(timeNowFn())
+		return gang.WaitTime, Wait
+	}
+	return 0, Success
+}
+
+// ActivateGang
+// Put all the pods belong to the Gang which in UnSchedulableQueue or backoffQueue back to activeQueue,
+func (p *GangPlugin) ActivateGang(pod *corev1.Pod, state *framework.CycleState) {
+	gangId := getNamespaceSplicingName(pod.Namespace,
+		getGangNameByPod(pod))
+
+	pods, err := p.podLister.Pods(pod.Namespace).List(labels.NewSelector())
+	if err != nil {
+		klog.Errorf("ActivateGang Failed to list pods belong to a Gang: %v", gangId)
+		return
+	}
+	toActivePods := make([]*corev1.Pod, 0)
+	for i := range pods {
+		gangIdTmp := getNamespaceSplicingName(pods[i].Namespace,
+			getGangNameByPod(pods[i]))
+		if gangIdTmp == gangId {
+			if pods[i].UID == pod.UID {
+				continue
+			}
+			toActivePods = append(toActivePods, pods[i])
+		}
+	}
+
+	if len(toActivePods) != 0 {
+		if c, err := state.Read(framework.PodsToActivateKey); err == nil {
+			if s, ok := c.(*framework.PodsToActivate); ok {
+				s.Lock()
+				for _, pod := range toActivePods {
+					namespacedName := getNamespaceSplicingName(pod.Namespace, pod.Name)
+					s.Map[namespacedName] = pod
+				}
+				s.Unlock()
+			}
+		}
+	}
+}
+
+func (p *GangPlugin) AllowGangGroup(pod *corev1.Pod) {
```

Comment:
it is better to use gang as the parameter, rather than pod.

Reply:
during the new implement [510](https://github.com/koordinator-sh/koordinator/pull/510), AlloGangGroup is the interface by PodGroupMagager, and it can only be start by the Coescheduling Plugin outside, which only has the parameter pod, so i think it's better to pass pod to the func